File: Source/Entities/SpeedcheckGate.cs
using Microsoft.Xna.Framework;
using Monocle;
using Celeste.Mod.Entities;
using Celeste;
using Celeste.Mod.ScugHelper;
using MonoMod.Cil;
using System;
using Celeste.Mod;
namespace Celeste.Mod.ScugHelper.Entities;

[CustomEntity("ScugHelper/SpeedcheckGate")]
public class SpeedcheckGate(EntityData data, Vector2 offset) : AbstractGate(data, offset)
{
    public enum SpeedAxis {
        Horizontal,
        Vertical,
        Total
    }
    public enum ComparisonType {
        Less,
        LessEq,
        Greater,
        GreaterEq
    }
    public enum TriggerAction {
        Kill,
        SetFlag,
        SetSpeed
    }
    
    public SpeedAxis Axis = data.Enum<SpeedAxis>("Axis");
    public ComparisonType Comparison = data.Enum<ComparisonType>("Comparison");
    public TriggerAction Action = data.Enum<TriggerAction>("Action");
    public float Threshold = data.Float("Threshold");

    public string FlagName = data.String("FlagName");
    public Vector2 SetSpeed = new(data.Float("SetX"), data.Float("SetY"));
    public bool ChangeX = data.Bool("ChangeX", true);
    public bool ChangeY = data.Bool("ChangeY", true);

    public override void OnTrigger(Player player) {
        float checkSpeed = Axis switch {
            SpeedAxis.Horizontal => Math.Abs(player.Speed.X),
            SpeedAxis.Vertical => Math.Abs(player.Speed.Y),
            SpeedAxis.Total => player.Speed.Length(),
        };
        bool isTriggered = Comparison switch {
            ComparisonType.Less => checkSpeed < Threshold,
            ComparisonType.LessEq => checkSpeed <= Threshold,
            ComparisonType.Greater => checkSpeed > Threshold,
            ComparisonType.GreaterEq => checkSpeed >= Threshold
        };
        if (!isTriggered) return;
        switch (Action) {
            case TriggerAction.Kill:
                player.Die(Vector2.Zero);
                break;
            case TriggerAction.SetFlag:
                SceneAs<Level>().Session.SetFlag(FlagName);
                break;
            case TriggerAction.SetSpeed:
                if (ChangeX) { player.X = MathF.Round(X + SetSpeed.X * Engine.DeltaTime / 2); player.Speed.X = SetSpeed.X; }
                if (ChangeY) { player.Y = MathF.Round(Y + SetSpeed.Y * Engine.DeltaTime / 2); player.Speed.Y = SetSpeed.Y; }
                break;
        }
    }
}

File: Loenn/entities/vol1/speedcheck_gate.lua
local utils = require("utils")
local drawing = require("utils.drawing")
local scughelper = require("mods").requireFromPlugin("libraries.scughelper")

return {
    name = "ScugHelper/SpeedcheckGate",
    depth = 100,
    placements = {
        {
            name = "normal",
            data = {
                Angle = 0,
                Size = 32,
                Threshold = 240,
                Axis = 0,
                Comparison = 0,
                Action = 0,
                FlagName = "",
                SetX = 0,
                SetY = 0,
                ChangeX = true,
                ChangeY = true,
            }
        },
    },
    sprite = function(room, entity)
        return scughelper.drawableGate(entity.x, entity.y, entity.Angle, entity.Size, {0, 1, 1})
    end,
    selection = function(room, entity)
        return utils.rectangle(
            entity.x - 8,
            entity.y - 8,
            16, 16
        )
    end,
    fieldInformation = {
        Axis = {
            type = "enum",
            options = { {"Horizontal", 0}, {"Vertical", 1}, {"Total", 2} }
        },
        Comparison = {
            type = "enum",
            options = { {"<", 0}, {"<=", 1}, {">", 2}, {">=", 3} }
        },
        Action = {
            type = "enum",
            options = { {"Kill", 0}, {"SetFlag", 1}, {"SetSpeed", 2} }
        }
    }
}
